bicytok/distance_metric_funcs.py

```diff
@@ -96,47 +117,142 @@ def KL_EMD_2D(
     assert targNorms.shape[0] != recAbundances.shape[0]
 
     row, col = np.tril_indices(
-        recAbundances.shape[1]
-    )  # Triangle indices, includes diagonal (k=0 by default)
+        recAbundances.shape[1], k=0
+    )  # k=0 includes diagonals which are 1D distances
     for rec1, rec2 in zip(row, col, strict=False):
+        if not calc_1D and rec1 == rec2:
+            continue
+
         targAbun1, targAbun2 = targNorms[:, rec1], targNorms[:, rec2]
         offTargAbun1, offTargAbun2 = offTargNorms[:, rec1], offTargNorms[:, rec2]
 
-        assert np.allclose(
-            targAbun1, recAbundances[targ, rec1] / np.mean(recAbundances[:, rec1])
-        )
+        if (
+            np.mean(recAbundances[:, rec1]) > 5
+            and np.mean(recAbundances[:, rec2]) > 5
+            and np.mean(targAbun1) > np.mean(offTargAbun1)
+            and np.mean(targAbun2) > np.mean(offTargAbun2)
+        ):
+            assert np.allclose(
+                targAbun1, recAbundances[targ, rec1] / np.mean(recAbundances[:, rec1])
+            )
+
+            targAbunAll = np.vstack((targAbun1, targAbun2)).transpose()
+            offTargAbunAll = np.vstack((offTargAbun1, offTargAbun2)).transpose()
+
+            assert targAbunAll.shape == (np.sum(targ), 2)
+
+            # Estimate the 2D probability distributions of the two receptors
+            targKDE = KernelDensity(kernel="gaussian").fit(targAbunAll)
```

Comment:
There is some repeated code in the body here that might make a good function to share across the 2D and 3D cases. Not here, but maybe note in an issue.